### PR TITLE
Replica sets calculation

### LINKED ISSUE #145: Calculate set of replicas for given token
In current implementation of token aware load balancing, we try only the owner of the range (vnode) that the token lies in. We should calculate the set of replicas for this token, using the replication strategy of the table being queried.

### PR DESCRIPTION
**Pull request by havaker**
## Description
Added replica sets calculation for tables using Simple Strategy or Network Topology Strategy.
Split load balancing module into submodules, each corresponding to one policy.
Refactored old tests and added new ones.

## Design
Replica sets are calculated each time some query is executed using Token-aware load balancing policy (it might be a good idea to cache/precompute them in the future).

## Fixed issues
Fixes #145

## Pre-review checklist

- [x]  I have split my patch into logically separate commits.
- [x] All commit messages clearly explain what they change and why.
- [x] I added relevant tests for new features and bug fixes.
- [x] All commits compile, pass static checks and pass test.
- [x] PR description sums up the changes and reasons why they should be introduced.
- [x] I added appropriate `Fixes:` annotations to PR description.

### CODE REVIEW

**Comment by havaker** [comment]
@psarna @piodul please review

File: scylla/src/transport/load_balancing.rs [hunk 1]
```diff
@@ -210,26 +206,26 @@ impl LoadBalancingPolicy for TokenAwarePolicy {
...
                     _ => {
                         // default to simple strategy with replication factor = 1
                         let replication_factor = 1;
-                        let mut replica =
+                        let replica =
                             Self::simple_strategy_replicas(cluster, &token, replication_factor);
-                        self.child_policy.apply_child_policy(&mut replica)
+                        self.child_policy.apply_child_policy(replica)
                     }
                 }
```
Comment (piodul) [inline scylla/src/transport/load_balancing.rs]: I think this whole `match` can be simplified:

```rust
let replicas = match strategy {
    Strategy::SimpleStrategy { replication_factor } => {
            Self::simple_strategy_replicas(cluster, &token, *replication_factor)
    }
    Strategy::NetworkTopologyStrategy {
        datacenter_repfactors,
    } => {
        Self::network_topology_strategy_replicas(
            cluster,
            &token,
            datacenter_repfactors,
        )
    }
    _ => {
        // default to simple strategy with replication factor = 1
        let replication_factor = 1;
        Self::simple_strategy_replicas(cluster, &token, replication_factor)
    }
};

self.child_policy.apply_child_policy(replicas)
```

**Review by piodul — COMMENTED** [review]
Left some minor comments. It also needs rebase. Apart from that, looks good.

I've noticed that the gocql driver precomputes replicas for each (keyspace x vnode) pair. I think our driver should do this, too, but the current approach is fine for now. We can tackle this in a separate issue/PR.

File: scylla/src/transport/load_balancing.rs (hunk 1 above)
Comment (havaker) [reply scylla/src/transport/load_balancing.rs]: Done

File: scylla/src/transport/load_balancing.rs [hunk 2]
```diff
@@ -114,20 +125,33 @@ impl LoadBalancingPolicy for TokenAwarePolicy {
...
+                        self.child_policy.apply_child_policy(&mut replicas)
+                    }
+                    _ => {
+                        // default to simple strategy with replication factor = 1
+                        let replication_factor = 1;
+                        let mut replica =
+                            Self::simple_strategy_replicas(cluster, &token, replication_factor);
+                        self.child_policy.apply_child_policy(&mut replica)
+                    }
+                }
```
[the remark this answers is not in the document]
Comment (havaker) [reply scylla/src/transport/load_balancing.rs]: Done

File: scylla/src/transport/load_balancing.rs [hunk 3]
```diff
@@ -114,20 +125,33 @@ impl LoadBalancingPolicy for TokenAwarePolicy {
...
-                    .map(|(_token, node)| node.clone())
-                    .unwrap_or_else(get_first_node);
-
-                let mut plan = std::iter::once(owner);
-                self.child_policy.apply_child_policy(&mut plan)
+                let keyspace = statement
+                    .keyspace
+                    .as_ref()
+                    .map(|k| cluster.keyspaces.get(k))
+                    .flatten();
```
[the remark this answers is not in the document]
Comment (havaker) [reply scylla/src/transport/load_balancing.rs]: Done

File: scylla/src/transport/load_balancing.rs [hunk 4]
```diff
@@ -111,10 +112,79 @@ impl TokenAwarePolicy {
...
+            .map(|(dc_name, repfactor)| {
+                let rack_count = cluster
+                    .datacenters
+                    .get(dc_name)
+                    .map(|dc| dc.rack_count)
+                    .unwrap_or(0);
+
+                (dc_name.clone(), repfactor - rack_count)
+            })
+            .collect::<HashMap<String, usize>>();
```
[the remark this answers is not in the document]
Comment (havaker) [reply scylla/src/transport/load_balancing.rs]: Done

File: scylla/src/statement/prepared_statement.rs [hunk 5]
```diff
@@ -114,6 +114,14 @@ impl PreparedStatement {
 
         Ok(buf.into())
     }
+
+    /// Returns the name of the keyspace this statement is operating on.
+    pub fn get_keyspace_name(&self) -> Option<String> {
```
[the remark this answers is not in the document]
Comment (havaker) [reply scylla/src/statement/prepared_statement.rs]: Done

**Comment by havaker** [comment]
Rebased

**Pull request merged**